accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+    ```
+
+*   `await for` is used when the function immediately enclosing the collection
+    literal is not asynchronous.
+
+    ```dart
+    main() {
+      [await for (_ in stream)]; // Error.
+    }
+    ```
+
+*   `await` is used before a C-style `for` element. `await` can only be used
+    with `for-in` loops.
+
+    ```dart
+    main() {
+      [await for (;;)]; // Error.
+    }
+    ```
+
+*   The type of the condition expression (the second clause) in a C-style `for`
+    element may not be assigned to `bool`.
+
+    ```dart
+    [for (; "not bool";) 1] // Error.
+    ```
+
+## Constant Semantics
+
+The runtime semantics below can also be used to determine the compile-time value
+of a collection literal marked `const`, with a few modifications:
+
+*   A `listLiteral` is constant if it occurs in a constant context or if it is
+    directly prefixed by `const`. If so, then its `elements` must all be
+    constant elements, or at least potentially constant elements.
+
+*   A `setOrMapLiteral` is constant if it occurs in a constant context or if it
+    is directly prefixed by `const`. If so, then its `elements` must all be
+    constant elements, or at least potentially constant elements.
+
+*   An `expressionElement` is a constant element if its expression is a constant
+    expression, and a potentially constant element if it's expression is a
+    potentially constant expression.
+
+*   A `mapEntry` is a constant element if both its key and value expression are
+    constant expressions, and a potentially constant element if both are
+    potentially constant expressions.
+
+*   A `spreadElement` is a constant element if its expression is constant and it
+    evaluates to a constant `List`, `Set` or `Map` instance originally created
+    by a list, set or map literal. It is a potentially constant element if the
+    expression is potentially constant expression.
+
+*   An `ifElement` is constant if its condition is a constant expression
+    evaluating to a Boolean value and either:
+
+    *   If the condition evaluates to `true`, the then element is a constant
+        expression, and any else element is a potentially constant element.
+
+    *   If the condition evaluates to `false`, then the then element is a
+        potentially constant element and any else element is a constant element.
+
+*   A `forElement` is never a constant element. `for` cannot be used in constant
+    set or map literals.
+
+*   It is a compile-time error to have duplicate values in a set literal or
+    duplicate keys in a map literal, according to behavior specified in the
+    runtime semantics.
+
+*   When evaluating a const collection literal, if there is an existing const
+    list, map, or set literal that contains the same series of values or entries
+    in the same order, then that object is used instead of producing a new one.
+    *In other words, constants are canonicalized.* Note that maps are only
```

Comment:
Maybe 'Note that constant maps and sets are distinct even after canonicalization if they do not agree on the ordering'.

Reply:
Tweaked to mention sets.